spec/models/miq_request_workflow_spec.rb

```diff
@@ -434,6 +434,26 @@
     end
   end
 
+  context "#set_request_values" do
+    it "doesn't reset owner_group and requester_group on a second run" do
+      owner = FactoryGirl.create(:user_with_email, :miq_groups => [FactoryGirl.create(:miq_group)])
+      values = {:owner_email => owner.email}
+      workflow.set_request_values(values)
+      expect(values[:owner_group]).to eq(owner.current_group.description)
+      expect(values[:requester_group]).to eq(workflow.requester.miq_group_description)
```

Comment:
I would suggest breaking this into 2 tests.  The first testing would be these `expect` calls to validate that `set_request_values` is populating the values.  A second test would reuse the workflow setup of the first and then create the new requester/owner and call `set_request_values` and validate the results.

The basic outline would be something like this:
```ruby
context "#set_request_values" do
  let(:workflow) do
    <workflow_setup>
  end

  it `sets requester/owner` do
    expect...
  end

  it 'does not reset requester/owner' do
    old_requester = workflow.requester
    ...
    workflow.set_request_values(values)
    expect...
  end
end
```

Reply:
Seems like a good idea, I'll revise.

Reply:
Actually, I'm not sure that works? The thing I need to share with persistence between tests is the value hash, which I can't do with `let` as I understand it. So even if I split this up the second test is still going to have to set up the value hash as though `set_request_values` has been called in order to demonstrate it's not being altered.

Reply:
How about this:

```ruby
context "#set_request_values" do
  before do
    workflow.set_request_values(values)
  end
  let(:values) { {:owner_email => owner.email} }
  let(:owner)  { FactoryGirl.create(:user_with_email, :miq_groups => [FactoryGirl.create(:miq_group)]) }

  it 'sets owner_group and requester_group' do
    expect(values[:owner_group]).to eq(owner.current_group.description)
    expect(values[:requester_group]).to eq(workflow.requester.miq_group_description)
  end

  it 'doesn't reset owner_group and requester_group on a second run' do
    old_requester = workflow.requester
    new_requester = FactoryGirl.create(:user_with_email, :miq_groups => [FactoryGirl.create(:miq_group)])
    workflow.requester = new_requester
    new_owner = FactoryGirl.create(:user_with_email, :miq_groups => [FactoryGirl.create(:miq_group)])

    values[:owner_email] = new_owner.email
    workflow.set_request_values(values)
    expect(values[:owner_group]).to eq(owner.current_group.description)
    expect(values[:requester_group]).to eq(old_requester.miq_group_description)
  end
end
```

Reply:
That does work. I'm slightly surprised at the way `let` and `before` interact, but now I know for the future. Thanks! :)